src/posix/platform/settings.cpp

```diff
@@ -32,429 +32,442 @@
+    } while (false)
 
-#ifdef __cplusplus
-extern "C" {
-#endif
+static const size_t kMaxFileNameSize = sizeof(OPENTHREAD_CONFIG_POSIX_SETTINGS_PATH) + 32;
 
-enum
-{
-    kBlockAddBeginFlag    = 0x1,
-    kBlockAddCompleteFlag = 0x02,
-    kBlockDeleteFlag      = 0x04,
-    kBlockIndex0Flag      = 0x08,
-};
+static int sFd = -1;
 
-enum
+static void getSettingsFileName(char aFileName[kMaxFileNameSize], bool aSwap)
 {
-    kSettingsFlagSize      = 4,
-    kSettingsBlockDataSize = 255,
+    const char *offset = getenv("PORT_OFFSET");
 
-    kSettingsInSwap = 0xbe5cc5ef,
-    kSettingsInUse  = 0xbe5cc5ee,
-    kSettingsNotUse = 0xbe5cc5ec,
-};
+    snprintf(aFileName, kMaxFileNameSize, OPENTHREAD_CONFIG_POSIX_SETTINGS_PATH "/%s_%" PRIx64 ".%s",
+             offset == NULL ? "0" : offset, NODE_ID, (aSwap ? "swap" : "data"));
+}
 
-OT_TOOL_PACKED_BEGIN
-struct settingsBlock
+static int swapOpen(void)
 {
-    uint16_t key;
-    uint16_t flag;
-    uint16_t length;
-    uint16_t reserved;
-} OT_TOOL_PACKED_END;
+    char fileName[kMaxFileNameSize];
+    int  fd;
 
-/**
- * @def SETTINGS_CONFIG_BASE_ADDRESS
- *
- * The base address of settings.
- *
- */
-#ifndef SETTINGS_CONFIG_BASE_ADDRESS
-#define SETTINGS_CONFIG_BASE_ADDRESS 0x39000
-#endif // SETTINGS_CONFIG_BASE_ADDRESS
+    getSettingsFileName(fileName, true);
+    fd = open(fileName, O_RDWR | O_CREAT | O_TRUNC, 0600);
+    VerifyOrDie(fd != -1);
 
-/**
- * @def SETTINGS_CONFIG_PAGE_SIZE
- *
- * The page size of settings.
- *
- */
-#ifndef SETTINGS_CONFIG_PAGE_SIZE
-#define SETTINGS_CONFIG_PAGE_SIZE 0x800
-#endif // SETTINGS_CONFIG_PAGE_SIZE
+    return fd;
+}
 
 /**
- * @def SETTINGS_CONFIG_PAGE_NUM
+ * This function reads @p aLength bytes from the data file and appends to the swap file.
  *
- * The page number of settings.
+ * @param[in]   aFd     The file descriptor of the current swap file.
+ * @param[in]   aLength Number of bytes to copy.
  *
  */
-#ifndef SETTINGS_CONFIG_PAGE_NUM
-#define SETTINGS_CONFIG_PAGE_NUM 2
-#endif // SETTINGS_CONFIG_PAGE_NUM
+static void swapWrite(int aFd, uint16_t aLength)
+{
+    const size_t kBlockSize = 512;
+    uint8_t      buffer[kBlockSize];
 
-static uint32_t sSettingsBaseAddress;
-static uint32_t sSettingsUsedSize;
+    while (aLength > 0)
+    {
+        uint16_t count = aLength >= sizeof(buffer) ? sizeof(buffer) : aLength;
+        ssize_t  rval  = read(sFd, buffer, count);
+
+        VerifyOrDie(rval > 0);
+        count = static_cast<uint16_t>(rval);
+        rval  = write(aFd, buffer, count);
+        assert(rval == count);
+        VerifyOrDie(rval == count);
+        aLength -= count;
+    }
+}
 
-static uint16_t getAlignLength(uint16_t length)
+static void swapPersist(int aFd)
 {
-    return (length + 3) & 0xfffc;
+    char swapFile[kMaxFileNameSize];
+    char dataFile[kMaxFileNameSize];
+
+    getSettingsFileName(swapFile, true);
+    getSettingsFileName(dataFile, false);
+
+    VerifyOrDie(0 == rename(swapFile, dataFile));
```

Comment:
`rename()` is ensured to be atomic, is this correct? 
I assume we cannot loose both the original dataFile and swapFile, if the device happens to crash in middle of `rename()`.

Reply:
Yes, `rename()` is an atomic operation. However, I think we should have an `fsync()` call after `rename()` to ensure that the change is committed to disk before moving forward. While `rename()` is atomic, it is possible to roll back to the old version if the system crashes before it is committed to disk.

Reply:
I added calls to `fsync()` and `sync()`.